### PR TITLE
Use little-endian encoding for Blake2 hashing on all architectures

### LINKED ISSUE #38891: ppc64/s390x `assertion failed: eps.windows(2).all(|w| w[0].cmp(self, &w[1]) != Ordering::Greater)`
While looking into #37996 on Fedora 25 ppc64, I hit a rustc assertion building rust-encoding.  The same nightly version is fine on Fedora 25 x86_64.  Stable and beta ppc64 are also fine, barring the linker problems from that former issue.

```
# rustc +nightly -Vv
rustc 1.16.0-nightly (6f1ae663e 2017-01-06)
binary: rustc
commit-hash: 6f1ae663ef21602841173c0b9549347904e3c9a8
commit-date: 2017-01-06
host: powerpc64-unknown-linux-gnu
release: 1.16.0-nightly
LLVM version: 3.9
```

```
# cargo +nightly build
   Compiling encoding_index_tests v0.1.5 (file:///tmp/rust-encoding/src/index/tests)
   Compiling encoding-types v0.2.0 (file:///tmp/rust-encoding/src/types)
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

thread 'rustc' panicked at 'assertion failed: eps.windows(2).all(|w| w[0].cmp(self, &w[1]) != Ordering::Greater)', /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:1467
note: Run with `RUST_BACKTRACE=1` for a backtrace.

Build failed, waiting for other jobs to finish...
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

thread 'rustc' panicked at 'assertion failed: eps.windows(2).all(|w| w[0].cmp(self, &w[1]) != Ordering::Greater)', /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:1467
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: Could not compile `encoding_index_tests`.

To learn more, run the command again with --verbose.
```

Here's the backtrace from the first occurrence:

```
thread 'rustc' panicked at 'assertion failed: eps.windows(2).all(|w| w[0].cmp(self, &w[1]) != Ordering::Greater)', /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:1467
stack backtrace:
   1:     0x3fff8f6ef8cf - rust_metadata_std_73a7c37de75e9a9d
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:42
   2:     0x3fff8f7011b3 - rust_metadata_std_73a7c37de75e9a9d
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libstd/panicking.rs:351
   3:     0x3fff8f700a7f - rust_metadata_std_73a7c37de75e9a9d
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libstd/panicking.rs:361
   4:     0x3fff8f7018f7 - rust_metadata_std_73a7c37de75e9a9d
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libstd/panicking.rs:555
   5:     0x3fff8e487bbf - rust_metadata_rustc_6cd08641c2891733
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libstd/panicking.rs:517
   6:     0x3fff8e6c2ce7 - rust_metadata_rustc_6cd08641c2891733
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:1467
   7:     0x3fff8ea0ab43 - rust_metadata_rustc_metadata_bc61964981617275
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:1501
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:1563
   8:     0x3fff8ea85be7 - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:1544
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:1501
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:437
   9:     0x3fff8ea83933 - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:779
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:206
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:486
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:200
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:485
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:180
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:150
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:176
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:92
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:169
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:92
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:375
  10:     0x3fff8ea833ef - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:779
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:180
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:371
  11:     0x3fff8ea578cf - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:779
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:206
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:38
  12:     0x3fff8ea84197 - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:200
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:36
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:180
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:138
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:176
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:92
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:169
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:92
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:375
  13:     0x3fff8ea833ef - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:779
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:180
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:371
  14:     0x3fff8ea5a7b7 - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:779
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/subst.rs:155
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:176
  15:     0x3fff8ea0c6f7 - rust_metadata_rustc_metadata_bc61964981617275
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/subst.rs:153
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:169
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/subst.rs:152
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:404
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/ops.rs:2706
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/option.rs:383
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/iter/mod.rs:999
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/result.rs:1072
  16:     0x3fff8ea0ba8f - rust_metadata_rustc_metadata_bc61964981617275
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/iter/iterator.rs:2178
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_data_structures/array_vec.rs:189
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_data_structures/accumulate_vec.rs:114
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/result.rs:1089
  17:     0x3fff8ea0ac4f - rust_metadata_rustc_metadata_bc61964981617275
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/iter/iterator.rs:1218
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:1563
  18:     0x3fff8ea850cf - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:1544
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:1511
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:404
  19:     0x3fff8ea841e7 - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:779
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:180
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:116
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:176
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:92
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:169
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:92
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:375
  20:     0x3fff8ea0c613 - rust_metadata_rustc_metadata_bc61964981617275
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:779
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:416
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/ops.rs:2706
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/option.rs:383
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/iter/mod.rs:999
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/result.rs:1072
  21:     0x3fff8ea0b853 - rust_metadata_rustc_metadata_bc61964981617275
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/iter/iterator.rs:2178
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_data_structures/array_vec.rs:189
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_data_structures/accumulate_vec.rs:114
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/result.rs:1089
  22:     0x3fff8ea0ae67 - rust_metadata_rustc_metadata_bc61964981617275
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/iter/iterator.rs:1218
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:1563
  23:     0x3fff8ea8583f - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:1544
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:1506
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:416
  24:     0x3fff8ea516c3 - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:779
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:206
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:567
  25:     0x3fff8ea526ef - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:200
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:565
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:206
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:486
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:200
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:485
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:206
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:547
  26:     0x3fff8ea858eb - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:200
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:542
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:422
  27:     0x3fff8ea8445f - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:779
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:180
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:142
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:176
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:92
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:169
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/sty.rs:92
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:375
  28:     0x3fff8ea8928b - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libserialize/serialize.rs:779
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:155
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/decoder.rs:665
  29:     0x3fff8ea95fff - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_metadata/cstore_impl.rs:89
  30:     0x3fff8e6d7327 - rust_metadata_rustc_6cd08641c2891733
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/mod.rs:2265
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/mod.rs:1882
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/dep_graph/dep_tracking_map.rs:145
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/mod.rs:1878
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/mod.rs:2263
  31:     0x3fff8f25d417 - rust_metadata_rustc_trans_f329661d828bb133
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_trans/back/symbol_names.rs:230
  32:     0x3fff8f25bf63 - rust_metadata_rustc_trans_f329661d828bb133
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_trans/back/symbol_export.rs:95
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/ops.rs:2706
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/option.rs:383
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/iter/mod.rs:999
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcollections/vec.rs:1632
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcollections/vec.rs:1615
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcollections/vec.rs:1502
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libcore/iter/iterator.rs:1218
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_trans/back/symbol_export.rs:89
  33:     0x3fff8f27dd43 - rust_metadata_rustc_trans_f329661d828bb133
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_trans/base.rs:1274
  34:     0x3fff8f8dacab - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_driver/driver.rs:1046
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/util/common.rs:48
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_driver/driver.rs:1044
  35:     0x3fff8f8b201f - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_driver/driver.rs:200
  36:     0x3fff8f8d9393 - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_driver/driver.rs:992
  37:     0x3fff8f832bef - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:974
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libstd/thread/local.rs:253
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:971
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:958
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libstd/thread/local.rs:253
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:955
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc/ty/context.rs:733
  38:     0x3fff8f8b02f7 - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_driver/driver.rs:867
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_driver/driver.rs:166
  39:     0x3fff8f8f1e4b - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_driver/lib.rs:221
  40:     0x3fff8f7f38a7 - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_driver/lib.rs:1119
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_driver/lib.rs:137
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/librustc_driver/lib.rs:1053
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libstd/panic.rs:296
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libstd/panicking.rs:460
  41:     0x3fff8f70cbff - rust_metadata_std_73a7c37de75e9a9d
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libpanic_unwind/lib.rs:98
  42:     0x3fff8f823447 - <unknown>
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libstd/panicking.rs:436
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libstd/panic.rs:361
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libstd/thread/mod.rs:357
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/liballoc/boxed.rs:605
  43:     0x3fff8f6ff767 - rust_metadata_std_73a7c37de75e9a9d
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/liballoc/boxed.rs:615
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libstd/sys_common/thread.rs:21
                        at /buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-rustbuild-linux/build/src/libstd/sys/unix/thread.rs:84
  44:     0x3fff8a71be63 - <unknown>
  45:     0x3fff8f541693 - <unknown>
```

### PR DESCRIPTION
**Pull request by arielb1**
Like many hash functions, the blake2 hash is mathematically defined on
a sequence of 64-bit words. As Rust's hash interface operates on
sequences of octets, some encoding must be used to bridge that
difference.

The Blake2 RFC (RFC 7693) specifies that:
```
   Byte (octet) streams are interpreted as words in little-endian order,
   with the least-significant byte first.
```

So use that encoding consistently.

Fixes #38891.

Beta-nominating since this is a regression since 1.15.

r? @michaelwoerister

### CODE REVIEW

File: src/librustc_data_structures/blake2b.rs [hunk 1]
```diff
@@ -113,17 +113,17 @@ fn blake2b_compress(ctx: &mut Blake2bCtx, last: bool) {
...
             let b: &mut [u8; 128] = &mut ctx.b;
             ::std::mem::transmute(b)
         };
 
-        // It's OK to modify the buffer in place since this is the last time
-        // this data will be accessed before it's overwritten
         if cfg!(target_endian = "big") {
             for word in &mut m[..] {
-                *word = word.to_be();
+                *word = u64::from_le(*word);
```
Comment (michaelwoerister) [inline src/librustc_data_structures/blake2b.rs]: Good catch!

File: src/librustc_data_structures/blake2b.rs [hunk 2]
```diff
@@ -113,17 +113,17 @@ fn blake2b_compress(ctx: &mut Blake2bCtx, last: bool) {
...
+        // Re-interpret the input buffer in the state as
+        // an array of big-endian u64s, converting them
+        // to machine endianness.
         let m: &mut [u64; 16] = unsafe {
             let b: &mut [u8; 128] = &mut ctx.b;
             ::std::mem::transmute(b)
         };
 
-        // It's OK to modify the buffer in place since this is the last time
-        // this data will be accessed before it's overwritten
```
Comment (michaelwoerister) [inline src/librustc_data_structures/blake2b.rs]: Why remove this comment?

File: src/librustc_data_structures/blake2b.rs [hunk 3]
```diff
@@ -113,17 +113,17 @@ fn blake2b_compress(ctx: &mut Blake2bCtx, last: bool) {
     }
 
     {
-        // Re-interpret the input buffer in the state as u64s
+        // Re-interpret the input buffer in the state as
+        // an array of big-endian u64s, converting them
+        // to machine endianness.
```
Comment (michaelwoerister) [inline src/librustc_data_structures/blake2b.rs]: Isn't it the other way round: We always make sure that this buffer is little-endian and now we are making sure to convert it to machine endianess?

**Review by michaelwoerister — COMMENTED** [review]
Thanks for the fix!
I think that one comment is should say little-endian, instead of big endian. And we probably should a comment that all data going in must be converted to little-endian (which the hasher should already do).

**Comment by arielb1** [comment]
Fixed nits.

> And we probably should a comment that all data going in must be converted to little-endian (which the hasher should already do).

Blake2b is a standard bytestream hasher. It should not do anything weird with its Hasher impl.

@bors r=michaelwoerister

**Comment by nikomatsakis** [comment]
@bors p=1

* Since this will likely go into beta, we need it to land ASAP

**Comment by nikomatsakis** [comment]
Accepting for beta. Small patch, regression. cc @rust-lang/compiler

**Pull request merged**